deployment/hasura/migrations/AerieMerlin/2_drop_profile_topic_fks/up.sql

```diff
@@ -0,0 +1,290 @@
+  'partitions concurrently with inserts to referenced tables, we have chosen to forego foreign keys from partitions'
+  'to other tables in favor of these hand-written triggers';
+
+create or replace function update_topic_cascade()
+  returns trigger
+  security invoker
+  language plpgsql as $$begin
+  if old.topic_index != new.topic_index or old.dataset_id != new.dataset_id
+  then
+    update event
+    set topic_index = new.topic_index,
+        dataset_id = new.dataset_id
+    where event.dataset_id = old.dataset_id and event.topic_index = old.topic_index;
+  end if;
+  return new;
+end$$;
+
+create trigger update_topic_trigger
+  after update on topic
+  for each row
+execute function update_topic_cascade();
+
+comment on trigger update_topic_trigger on topic is e''
+  'Trigger to simulate an ON UPDATE CASCADE foreign key constraint between event and topic. The reason to'
+  'implement this as a trigger is that this single trigger can propagate updates to any partitions of event.'
+  'If we used a foreign key, every new partition of event would need to add a new trigger to the topic'
+  'table - which requires acquiring a lock that conflicts with concurrent inserts. In order to allow adding new'
+  'partitions concurrently with inserts to referenced tables, we have chosen to forego foreign keys from partitions'
+  'to other tables in favor of these hand-written triggers';
+
+-- span
+alter table span drop constraint span_has_parent_span;
+alter table span drop constraint span_owned_by_dataset;
+
+create or replace function span_integrity_function()
+  returns trigger
+  security invoker
+  language plpgsql as $$begin
+  if not exists(select from dataset where dataset.id = new.dataset_id for key share of dataset)
+  then
+    raise exception 'foreign key violation: there is no dataset with id %', new.dataset_id;
+  end if;
+  return new;
+end$$;
+
+comment on function span_integrity_function is e''
+  'Used to simulate a foreign key constraint between span and dataset, to avoid acquiring a lock on the'
+  'dataset table when creating a new partition of span. This function checks that a corresponding dataset'
+  'exists for every inserted or updated span. A trigger that calls this function is added separately to each'
+  'new partition of span.';
+
+create constraint trigger insert_update_span_trigger
+  after insert or update on span
+  for each row
+execute function span_integrity_function();
+
+do $$
+  declare
+    dataset_id integer;
+  begin
+    for dataset_id in select id from dataset
+      loop
+        if exists(select from pg_tables where schemaname = 'public' and tablename = 'span_' || dataset_id) then
+          execute 'alter table span_' || dataset_id || ' add constraint span_has_parent_span foreign key (dataset_id, parent_id) references span_' || dataset_id || ' on update cascade on delete cascade;';
+        end if;
+      end loop;
+  end
+$$;
```

Comment:
Is this essentially running `span_add_foreign_key_to_partition` to all of the existing partition tables?

Reply:
It is indeed :+1: I've updated it to call that instead